include/cantera/numerics/FuncEval.h

```diff
@@ -49,6 +52,38 @@ class FuncEval
      */
     int eval_nothrow(double t, double* y, double* ydot);
 
+    //! Evaluate the setup processes for the Jacobian preconditioner.
+    //! @param[in] t time.
+    //! @param[in] y solution vector, length neq()
+    //! @param[out] ydot rate of change of solution vector, length neq()
+    //! @param[in] params sensitivity parameters
+    //! @param gamma the gamma in M=I-gamma*J
+    virtual void preconditionerSetup(double t, double* y, double* ydot, double gamma){
+        throw NotImplementedError("FuncEval::preconditionerSetup");
+    }
+
+    //! Evaluate the system using a Jacobian preconditioner. Called by the integrator.
+    //! @param[in] t time.
+    //! @param[in] y solution vector, length neq()
+    //! @param[out] ydot rate of change of solution vector, length neq()
+    //! @param[in] rhs right hand side vector used in linear system
+    //! @param[out] output guess vector used by GMRES
+    virtual void preconditionerSolve(double t, double* y, double* ydot, double* rhs, double* output){
+        throw NotImplementedError("FuncEval::preconditionerSolve");
+    }
+
+    //! preconditioner setup that doesn't throw an error but returns a
+    //! CVODES flag. It also helps as a first level of polymorphism
+    //! which identifies the specific FuncEval, e.g., ReactorNet.
+    //! Parameters are the same as preconditionerSetup
```

Comment:
```suggestion
    //! Parameters are the same as preconditionerSetup()
```
Ditto on `preconditionerSolve` below.